### PR TITLE
systemd: Don't open and close tuned DBus connections asynchronously

### PR DESCRIPTION
**Pull request by mvollmer**
The TunedPerformanceProfile component was keeping a DBus connection open to tuned, and was closing and reopening it whenever something might have changed that requires it.

This happened asynchronously with the rest of the code that was using this DBus connection, and would frequently cause that code to fail with "disconnected" errors.

Also, the TunedDialog was itself causing changes that require a new DBus connection (namely starting the tuned.service), but was always working with the one that was originally passed into it via its properties.

Let's straighten this all out:

 - Polling for the button state will open its own dedicated DBUs connection and close it when polling is done.  There is no need to keep a DBus connection open for the whole Cockpit session.

 - The tuned.service is started when the dialog is opened and a DBus
   connection is opened explicitly for the dialog after that and
   closed when the dialog is closed.

### CODE REVIEW

**Comment by mvollmer** [comment]
Starting the service might be slow and it would be best to do it inside the dialog, while the spinner is showing. That is probably straightforward.

Also, I think our DBusClient is supposed to survive the dbus service coming and going, maybe with some flags and owner watching. That would be another, even cleaner, way of handling this, by only having a single DBusClient that is never closed and always works.

File: pkg/systemd/overview-cards/tuned-dialog.jsx [hunk 1]
```diff
@@ -303,14 +300,14 @@ const TunedDialog = ({
                className="ct-m-stretch-body"
                isOpen
                help={help}
-               onClose={Dialogs.close}
+               onClose={() => dialogResult.resolve()}
```
Comment (jelly) [inline pkg/systemd/overview-cards/tuned-dialog.jsx:303]: The equivalent would be dialogResult.resolve

**Review by jelly — COMMENTED** [review]
Tested locally, the dialog loads **a lot** faster!

File: pkg/systemd/overview-cards/tuned-dialog.jsx (hunk 1 above)
Comment (mvollmer) [reply pkg/systemd/overview-cards/tuned-dialog.jsx:303]: Not necessarily. The `this` inside the call is different, and that might matter.

File: pkg/systemd/overview-cards/tuned-dialog.jsx [hunk 2]
```diff
@@ -31,84 +31,89 @@
...
+        tuned.call('/Tuned', 'com.redhat.tuned.control', 'active_profile', []),
+        tuned.call('/Tuned', 'com.redhat.tuned.control', 'recommend_profile', [])
+    ])
+        .then(([[is_running], [active_result], [recommended]]) => {
+            const active = is_running ? active_result : "none";
+            return ({ state: "running", active, recommended });
+        })
+        .catch((ex) => {
+            if (!tunedService.exists)
+                return ({ state: "not-installed" });
```
Comment (cockpituous) [inline pkg/systemd/overview-cards/tuned-dialog.jsx:51]: This added line is not executed by any test.

File: pkg/systemd/overview-cards/tuned-dialog.jsx [hunk 3]
```diff
@@ -31,84 +31,89 @@
...
+            else if (tunedService.state != "running")
+                return ({ state: "not-running" });
```
Comment (cockpituous) [inline pkg/systemd/overview-cards/tuned-dialog.jsx:53]: This added line is not executed by any test.

File: pkg/systemd/overview-cards/tuned-dialog.jsx [hunk 4]
```diff
@@ -31,84 +31,89 @@
...
+
+    const tunedService = useInit(() => service.proxy("tuned.service"));
+
+    async function update_with_tuned(tuned) {
+        try {
+            const { state, active, recommended } = await poll_tuned_state(tuned, tunedService);
+            let status;
+
+            if (state == "not-installed")
+                status = _("Tuned is not available");
```
Comment (cockpituous) [inline pkg/systemd/overview-cards/tuned-dialog.jsx:74]: This added line is not executed by any test.

File: pkg/systemd/overview-cards/tuned-dialog.jsx [hunk 5]
```diff
@@ -303,14 +300,14 @@
                className="ct-m-stretch-body"
                isOpen
                help={help}
-               onClose={Dialogs.close}
+               onClose={() => dialogResult.resolve()}
```
Comment (cockpituous) [inline pkg/systemd/overview-cards/tuned-dialog.jsx:303]: This added line is not executed by any test.

File: pkg/systemd/overview-cards/tuned-dialog.jsx [hunk 6]
```diff
@@ -303,14 +300,14 @@
...
                title={_("Change performance profile")}
                footer={
                    <>
                        <Button variant='primary' isDisabled={!selected} onClick={setProfile}>
                            {_("Change profile")}
                        </Button>
-                       <Button variant='link' onClick={Dialogs.close}>
+                       <Button variant='link' onClick={() => dialogResult.resolve()}>
```
Comment (cockpituous) [inline pkg/systemd/overview-cards/tuned-dialog.jsx:310]: This added line is not executed by any test.

File: pkg/systemd/overview-cards/tuned-dialog.jsx (hunk 2 above)
Comment (cockpituous) [inline pkg/systemd/overview-cards/tuned-dialog.jsx:51]: This added line is not executed by any test.

File: pkg/systemd/overview-cards/tuned-dialog.jsx (hunk 3 above)
Comment (cockpituous) [inline pkg/systemd/overview-cards/tuned-dialog.jsx:53]: This added line is not executed by any test.

File: pkg/systemd/overview-cards/tuned-dialog.jsx (hunk 4 above)
Comment (cockpituous) [inline pkg/systemd/overview-cards/tuned-dialog.jsx:74]: This added line is not executed by any test.

File: pkg/systemd/overview-cards/tuned-dialog.jsx (hunk 5 above)
Comment (cockpituous) [inline pkg/systemd/overview-cards/tuned-dialog.jsx:303]: This added line is not executed by any test.

File: pkg/systemd/overview-cards/tuned-dialog.jsx (hunk 6 above)
Comment (cockpituous) [inline pkg/systemd/overview-cards/tuned-dialog.jsx:310]: This added line is not executed by any test.

**Comment by mvollmer** [comment]
> Starting the service might be slow and it would be best to do it inside the dialog, while the spinner is showing. That is probably straightforward.
> 
> Also, I think our DBusClient is supposed to survive the dbus service coming and going, maybe with some flags and owner watching. That would be another, even cleaner, way of handling this, by only having a single DBusClient that is never closed and always works.

Done in #21115.

**Comment by mvollmer** [comment]
> Tested locally, the dialog loads **a lot** faster!

This is due to this: https://github.com/cockpit-project/cockpit/pull/21114/files#diff-edcd27aeefd91df68948709cd630aaefbc569ac0674cc54693139880404e17bdR101-R108

File: pkg/systemd/overview-cards/tuned-dialog.jsx [hunk 7]
```diff
@@ -31,84 +31,89 @@
...
+                status = _("Tuned is off");
+            else if (active == recommended)
+                status = _("This system is using the recommended profile");
+            else
+                status = _("This system is using a custom profile");
+            setBtnText(state == "running" ? active : _("none"));
+            setState(state);
+            setStatus(status);
+        } catch (ex) {
+            console.warn("failed to poll tuned", ex);
```
Comment (cockpituous) [inline pkg/systemd/overview-cards/tuned-dialog.jsx:87]: These 2 added lines are not executed by any test.

File: pkg/systemd/overview-cards/tuned-dialog.jsx [hunk 8]
```diff
@@ -31,84 +31,89 @@
...
+
+            setBtnText("error");
+            setStatus(_("Communication with tuned has failed"));
```
Comment (cockpituous) [inline pkg/systemd/overview-cards/tuned-dialog.jsx:90]: These 2 added lines are not executed by any test.

File: pkg/systemd/overview-cards/tuned-dialog.jsx (hunk 6 above)
Comment (cockpituous) [inline pkg/systemd/overview-cards/tuned-dialog.jsx:310]: This added line is not executed by any test.

File: pkg/systemd/overview-cards/tuned-dialog.jsx (hunk 1 above)
Comment (cockpituous) [inline pkg/systemd/overview-cards/tuned-dialog.jsx:303]: This added line is not executed by any test.

File: pkg/systemd/overview-cards/tuned-dialog.jsx [hunk 9]
```diff
@@ -303,14 +300,14 @@ const TunedDialog = ({
...
-               onClose={Dialogs.close}
+               onClose={() => dialogResult.resolve()}
                title={_("Change performance profile")}
                footer={
                    <>
                        <Button variant='primary' isDisabled={!selected} onClick={setProfile}>
                            {_("Change profile")}
                        </Button>
-                       <Button variant='link' onClick={Dialogs.close}>
+                       <Button variant='link' onClick={() => dialogResult.resolve()}>
```
Comment (cockpituous) [inline pkg/systemd/overview-cards/tuned-dialog.jsx:310]: This added line is not executed by any test.

**Comment by mvollmer** [comment]
> Starting the service might be slow and it would be best to do it inside the dialog, while the spinner is showing. That is probably straightforward.

@jelly, what do you think about this? Should we start tuned.service while the dialog is already open? Then the TunedDialog component needs to maintain its own DBusClient ins a `useState` hook, probably.

**Comment by jelly** [comment]
> > Starting the service might be slow and it would be best to do it inside the dialog, while the spinner is showing. That is probably straightforward.
> 
> @jelly, what do you think about this? Should we start tuned.service while the dialog is already open? Then the TunedDialog component needs to maintain its own DBusClient ins a `useState` hook, probably.

Hmmm right, what if it fails the error message get's "lost", so maybe we should do the extra work (and right thing) And start the service in the dialog?

**Comment by mvollmer** [comment]
> > > Starting the service might be slow and it would be best to do it inside the dialog, while the spinner is showing. That is probably straightforward.
> > 
> > 
> > @jelly, what do you think about this? Should we start tuned.service while the dialog is already open? Then the TunedDialog component needs to maintain its own DBusClient ins a `useState` hook, probably.
> 
> Hmmm right, what if it fails the error message get's "lost", so maybe we should do the extra work (and right thing) And start the service in the dialog?

Yes, catching the error is convincing.  I have done this now. The explicit `close()` function for closing the DBus channel is not great, that should rather be done in a "destroy" function of `useObject` or `useEffect`... do we care enough to refactor the code more?

**Comment by mvollmer** [comment]
> that should rather be done in a "destroy" function of `useObject` or `useEffect`... do we care enough to refactor the code more?

Yes, I did that as well.

**Review by jelly — APPROVED** [review]

**Pull request merged**